File: examples/relax_perl/relax.pl
#!/usr/bin/env perl

use strict;
use warnings;
use FindBin;
use File::Copy qw(copy);
use File::Spec;

use lib "$FindBin::Bin/../../src/httk/workflow/native/perl";
use HttkWorkflow;

my @collect = qw(INCAR KPOINTS OUTCAR CONTCAR OSZICAR vasprun.xml vasp-run-report.json);

sub env_or {
    my ($name, $fallback) = @_;
    return defined($ENV{$name}) && $ENV{$name} ne '' ? $ENV{$name} : $fallback;
}

sub stage_input {
    my ($attempt, $job_dir, $parameter, $fallback, $destination) = @_;
    my $relative = $attempt->parameter($parameter, $fallback);
    return -1 unless defined($relative);
    my $source = File::Spec->catfile($job_dir, $relative);
    return 0 unless -f $source;
    return copy($source, $destination) ? 1 : -1;
}

sub step_prepare {
    my ($attempt) = @_;
    my $job_dir = env_or('HTTK_WORKFLOW_JOB_DIR', '.');
    if (stage_input($attempt, $job_dir, 'poscar', 'files/POSCAR', 'POSCAR') <= 0) {
        $attempt->fail('vasp.input_missing', 'the starting structure is not in this payload', 0);
        return 0;
    }
    stage_input($attempt, $job_dir, 'incar', 'files/INCAR', 'INCAR');
    $attempt->runlog_note('prepared a relaxation');
    $attempt->advance('run', []);
    return 0;
}

sub step_run {
    my ($attempt) = @_;
    my $from_parameter = $attempt->parameter('vasp_command', '');
    my $command = $attempt->setting('vasp.command', $from_parameter);
    $command = '' unless defined($command);
    if ($command =~ /^\s*$/) {
        $attempt->fail(
            'vasp.command_missing',
            'no VASP command is configured: set it with '
                . "httk workflow workspace settings set vasp.command '...', or set "
                . 'HTTK_VASP_COMMAND, or give the job a vasp_command parameter',
            0,
        );
        return 0;
    }

    my $timeout = $attempt->parameter('timeout', '86400');
    my @tokens = split /\s+/, $command;
    my @args = ('--timeout', $timeout, '--report', 'vasp-run-report.json', '--', @tokens);
    my $status = $attempt->run(\@args);
    if ($status == 0) {
        $attempt->state_set('classification', 'completed');
        $attempt->runlog_note('VASP completed');
        $attempt->advance('publish', []);
    } else {
        $attempt->fail('vasp.failed', "VASP did not complete (status $status)", 0);
    }
    return 0;
}

sub step_publish {
    my ($attempt) = @_;
    my $prefix = $attempt->parameter('data_prefix', 'vasp');
    my $data_dir = env_or('HTTK_WORKFLOW_DATA_DIR', '');
    for my $name (@collect) {
        next unless -f $name;
        $attempt->put($name, "$prefix/$name") if $data_dir ne '';
    }
    $attempt->runlog_note($data_dir ne '' ? 'published to transactional data' : 'kept the result in the workdir');
    $attempt->succeed();
    return 0;
}

my $runner = HttkWorkflow::Runner->new(
    workflow => 'httk.vasp.relax-perl',
    steps => [qw(prepare run publish)],
);
$runner->step(prepare => \&step_prepare)
    ->step(run => \&step_run)
    ->step(publish => \&step_publish)
    ->main();

File: examples/relax_java/Relax.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.Arrays;

import static java.lang.System.getenv;

/** One minimal three-step VASP relaxation using the native Java SDK. */
public final class Relax {
    private static final String[] COLLECT = {
        "INCAR", "KPOINTS", "OUTCAR", "CONTCAR", "OSZICAR", "vasprun.xml", "vasp-run-report.json"
    };

    private Relax() {
    }

    private static String envOr(String name, String fallback) {
        String value = getenv(name);
        return value == null || value.isEmpty() ? fallback : value;
    }

    private static int stageInput(
            HttkWorkflow.Attempt attempt,
            String jobDir,
            String parameter,
            String fallback,
            String destination) throws IOException {
        String relative = attempt.parameter(parameter, fallback).orElse(fallback);
        Path source = Path.of(jobDir, relative);
        if (!Files.isRegularFile(source)) {
            return 0;
        }
        Files.copy(source, Path.of(destination), StandardCopyOption.REPLACE_EXISTING);
        return 1;
    }

    private static int prepare(HttkWorkflow.Attempt attempt) throws IOException {
        String jobDir = envOr("HTTK_WORKFLOW_JOB_DIR", ".");
        if (stageInput(attempt, jobDir, "poscar", "files/POSCAR", "POSCAR") <= 0) {
            attempt.fail("vasp.input_missing", "the starting structure is not in this payload", false);
            return 0;
        }
        stageInput(attempt, jobDir, "incar", "files/INCAR", "INCAR");
        attempt.runlogNote("prepared a relaxation");
        attempt.advance("run");
        return 0;
    }

    private static int run(HttkWorkflow.Attempt attempt) {
        String fromParameter = attempt.parameter("vasp_command", "").orElse("");
        String command = attempt.setting("vasp.command", fromParameter).orElse("");
        if (command.trim().isEmpty()) {
            attempt.fail(
                    "vasp.command_missing",
                    "no VASP command is configured: set it with "
                            + "httk workflow workspace settings set --key vasp.command --value '...' WORKSPACE, or set "
                            + "HTTK_VASP_COMMAND, or give the job a vasp_command parameter",
                    false);
            return 0;
        }

        String timeout = attempt.parameter("timeout", "86400").orElse("86400");
        String[] tokens = command.trim().split("\\s+");
        String[] args = new String[5 + tokens.length];
        args[0] = "--timeout";
        args[1] = timeout;
        args[2] = "--report";
        args[3] = "vasp-run-report.json";
        args[4] = "--";
        System.arraycopy(tokens, 0, args, 5, tokens.length);

        int status = attempt.run(args);
        if (status == 0) {
            attempt.stateSet("classification", "completed");
            attempt.runlogNote("VASP completed");
            attempt.advance("publish");
        } else {
            attempt.fail("vasp.failed", "VASP did not complete (status " + status + ")", false);
        }
        return 0;
    }

    private static int publish(HttkWorkflow.Attempt attempt) {
        String prefix = attempt.parameter("data_prefix", "vasp").orElse("vasp");
        String dataDir = envOr("HTTK_WORKFLOW_DATA_DIR", "");
        for (String name : COLLECT) {
            if (Files.isRegularFile(Path.of(name)) && !dataDir.isEmpty()) {
                attempt.put(name, prefix + "/" + name);
            }
        }
        attempt.runlogNote(dataDir.isEmpty() ? "kept the result in the workdir" : "published to transactional data");
        attempt.succeed();
        return 0;
    }

    public static void main(String[] args) {
        new HttkWorkflow.Runner("httk.vasp.relax-java", Arrays.asList("prepare", "run", "publish"))
                .step("prepare", Relax::prepare)
                .step("run", Relax::run)
                .step("publish", Relax::publish)
                .main(args);
    }
}
